File: 2023/JS/day-01.js
const fs = require("fs");
const path = require("path");

const data = fs.readFileSync(path.resolve(__dirname, "../inputs/day1.txt"), "utf-8")

const input = data.trim().split("\n");

//PART 1
let result = 0;
for (const line of input) {
  //remove all non digit characters in the input line
  const digits = line.replace(/\D/g, "");
  //add the first character to the last character per digit or to itself, in cases where the digit has just one character
  //convert to number
  //add each number to get our final result
  result += parseInt(digits[0] + (digits[digits.length - 1] ?? digits[0]));
}

console.log("RESULT: ", result);

//PART 2
let result2 = 0;
//create an object with keys = digits in words and the values = actual digits
const nums = {
  one: 1,
  two: 2,
  three: 3,
  four: 4,
  five: 5,
  six: 6,
  seven: 7,
  eight: 8,
  nine: 9,
};
for (const line of input) {
  const numbers = [];

  for (const [key, value] of Object.entries(nums)) {
    //this allows us get matches for `three` and `3` for instance when looping through our input.
    //returns something like this for each line of input:
    /**
     * [
          [ 'two', index: 9, input: 'nkxmdshm5twoseven672', groups: undefined ],
          [ '6', index: 14, input: '88788jnscmpqr66sxcjx', groups: undefined ],
          [ '7', index: 2, input: '88788jnscmpqr66sxcjx', groups: undefined ],
        ]
     */
    numbers.push(
      ...line.matchAll(new RegExp(key, "g")),
      ...line.matchAll(new RegExp(value, "g")),
    );
  }
  //this arranges the returned `numbers` based on their position in the string.
  //this allows for the manipulation when adding as shown below.
  numbers.sort((a, b) => a.index - b.index);

  //for numbers as words, the length would be greater than 1. In such a case, we use the respective value based on the nums object.
  //for regular numbers, the length would be 1, and they're simply cnnverted to integers
  //this is done for the first and last `digit` in each line of input
  //add each number to get our final result.
  result2 += parseInt(
    `${
      numbers[0][0].length > 1 ? nums[numbers[0][0]] : parseInt(numbers[0][0])
    }${
      numbers[numbers.length - 1][0].length > 1
        ? nums[numbers[numbers.length - 1][0]]
        : parseInt(numbers[numbers.length - 1][0])
    }`,
  );
}

console.log("RESULT 2: ", result2);

File: 2023/KT/src/Day01.kt
import java.io.File

//todd solution
fun main() {
    val d = Day01(File("../inputs/day01.txt").readLines())
    println(d.solvePart1())
    println(d.solvePart2())
}

class Day01(private val input: List<String>) {
    private val words: Map<String, Int> = mapOf(
        "one" to 1,
        "two" to 2,
        "three" to 3,
        "four" to 4,
        "five" to 5,
        "six" to 6,
        "seven" to 7,
        "eight" to 8,
        "nine" to 9
    )

    fun testPart2() = "four82nine74".possibleWordsAt(1)

    //adds all number values from `calibrationValue(it)`
    fun solvePart1(): Int = input.sumOf { calibrationValue(it) }

    //adds all number values in resulting list
    fun solvePart2(): Int = input.sumOf { row: String ->
        calibrationValue(
            //run through each char and turn into a digit or a null
            //and then map each of them to a string. in theory, we could
            //take the first and last digits from the resulting list instead of joining
            row.mapIndexedNotNull { index: Int, c ->
                //if digit, take as is
                if (c.isDigit()) c
                else
                //otherwise, see if this is the start of a word and if so
                //map to the digit that it represents.
                    row.possibleWordsAt(index).firstNotNullOfOrNull { candidate -> words[candidate] }
           //creates a list of strings separated by `,`
            }.joinToString()
        )
    }

    //check for the first and last instance of a digit character
    //concatenate them as strings
    //return them as numbers
    private fun calibrationValue(row: String): Int = "${row.first { it.isDigit() }}${row.last { it.isDigit() }}".toInt()

    //takes an integer as the starting point representing the index of the string
    //returns a list of 3 values
    //the values are substrings of the parsed string returning at most 3, 4 and 5 chars respectively
    private fun String.possibleWordsAt(startingAt: Int): List<String> = (3..5).map { len: Int ->
        substring(startingAt, (startingAt + len).coerceAtMost(length))
    }
}

// seb solution
//val words = listOf("one", "two", "three", "four", "five", "six", "seven", "eight", "nine")
//val revWords = words.map { it.reversed() }
//fun main() {
//    part1()
//    part2()
//}
//
//fun part1() {
//    val lines = readInput("day01")
//    println(lines.sumOf { line: String -> line.first { it.isDigit() }.digitToInt() * 10 + line.last { it.isDigit() }.digitToInt() })
//}
//
//fun part2() {
//    val lines = readInput("day01")
//    println(lines.sumOf { line -> getNumber(line, StartFrom.BEGINNING) * 10 + getNumber(line, StartFrom.END) })
//}
//
//enum class StartFrom { BEGINNING, END }
//
//private fun getNumber(line: String, startFrom: StartFrom): Int {
//    val indices = when (startFrom) {
//        StartFrom.BEGINNING -> line.indices
//        StartFrom.END -> line.lastIndex downTo 0
//    }
//    for (idx in indices) {
//        line[idx].digitToIntOrNull()?.let { return it }
//        for ((wordIndex, word) in words.withIndex()) {
//            if (line.substring(idx).startsWith(word)) {
//                return (wordIndex + 1)
//            }
//        }
//    }
//    error("Unreachable")
//}
